File: src/main/java/io/github/aalsanie/codes/protocol/http/HttpOutcomeMapper.java
package io.github.aalsanie.codes.protocol.http;

import io.github.aalsanie.codes.MappingResult;
import io.github.aalsanie.codes.OutcomeCode;
import io.github.aalsanie.codes.OutcomeDefinition;
import io.github.aalsanie.codes.OutcomeMapper;
import io.github.aalsanie.codes.StandardOutcomes;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

public final class HttpOutcomeMapper implements OutcomeMapper<HttpStatusCode> {
    private final Map<OutcomeCode, HttpStatusCode> mappings;

    private HttpOutcomeMapper(Map<OutcomeCode, HttpStatusCode> mappings) {
        this.mappings = Collections.unmodifiableMap(new LinkedHashMap<>(mappings));
    }

    public static HttpOutcomeMapper empty() {
        return new HttpOutcomeMapper(Map.of());
    }

    public static HttpOutcomeMapper standard() {
        LinkedHashMap<OutcomeCode, HttpStatusCode> values = new LinkedHashMap<>();
        values.put(StandardOutcomes.OK.getCode(), HttpStatusCode.OK);
        values.put(StandardOutcomes.INVALID_ARGUMENT.getCode(), HttpStatusCode.BAD_REQUEST);
        values.put(StandardOutcomes.UNAUTHENTICATED.getCode(), HttpStatusCode.UNAUTHORIZED);
        values.put(StandardOutcomes.PERMISSION_DENIED.getCode(), HttpStatusCode.FORBIDDEN);
        values.put(StandardOutcomes.NOT_FOUND.getCode(), HttpStatusCode.NOT_FOUND);
        values.put(StandardOutcomes.ALREADY_EXISTS.getCode(), HttpStatusCode.CONFLICT);
        values.put(StandardOutcomes.OUT_OF_RANGE.getCode(), HttpStatusCode.BAD_REQUEST);
        values.put(StandardOutcomes.RATE_LIMITED.getCode(), HttpStatusCode.TOO_MANY_REQUESTS);
        values.put(StandardOutcomes.UNIMPLEMENTED.getCode(), HttpStatusCode.NOT_IMPLEMENTED);
        values.put(StandardOutcomes.UNAVAILABLE.getCode(), HttpStatusCode.SERVICE_UNAVAILABLE);
        values.put(StandardOutcomes.INTERNAL.getCode(), HttpStatusCode.INTERNAL_SERVER_ERROR);
        values.put(StandardOutcomes.DATA_LOSS.getCode(), HttpStatusCode.INTERNAL_SERVER_ERROR);
        return new HttpOutcomeMapper(values);
    }

    @Override
    public MappingResult<HttpStatusCode> map(OutcomeDefinition definition) {
        Objects.requireNonNull(definition, "definition");
        HttpStatusCode status = mappings.get(definition.getCode());
        return status == null ? MappingResult.unmapped() : MappingResult.mapped(status);
    }

    public int getSize() {
        return mappings.size();
    }

    public boolean contains(OutcomeCode code) {
        return mappings.containsKey(Objects.requireNonNull(code, "code"));
    }

    public HttpOutcomeMapper withMapping(OutcomeDefinition definition, HttpStatusCode status) {
        Objects.requireNonNull(definition, "definition");
        Objects.requireNonNull(status, "status");
        OutcomeCode code = definition.getCode();
        if (mappings.containsKey(code)) {
            throw new IllegalArgumentException("mapping already exists for outcome code: " + code);
        }
        LinkedHashMap<OutcomeCode, HttpStatusCode> next = new LinkedHashMap<>(mappings);
        next.put(code, status);
        return new HttpOutcomeMapper(next);
    }

    public HttpOutcomeMapper withOverride(OutcomeDefinition definition, HttpStatusCode status) {
        Objects.requireNonNull(definition, "definition");
        Objects.requireNonNull(status, "status");
        OutcomeCode code = definition.getCode();
        if (!mappings.containsKey(code)) {
            throw new IllegalArgumentException("cannot override missing mapping for outcome code: " + code);
        }
        LinkedHashMap<OutcomeCode, HttpStatusCode> next = new LinkedHashMap<>(mappings);
        next.put(code, status);
        return new HttpOutcomeMapper(next);
    }

    public Map<OutcomeCode, HttpStatusCode> mappings() {
        return mappings;
    }
}

File: src/main/kotlin/io/github/aalsanie/codes/protocol/http/HttpOutcomeMapper.kt
package io.github.aalsanie.codes.protocol.http

import io.github.aalsanie.codes.InternalMappings
import io.github.aalsanie.codes.MappingResult
import io.github.aalsanie.codes.Outcome
import io.github.aalsanie.codes.OutcomeCode
import io.github.aalsanie.codes.OutcomeDefinition
import io.github.aalsanie.codes.OutcomeMapper
import io.github.aalsanie.codes.StandardOutcomes

class HttpOutcomeMapper private constructor(
    private val mappings: InternalMappings<HttpStatusCode>,
) : OutcomeMapper<HttpStatusCode> {
    val size: Int
        get() = mappings.size

    override fun map(definition: OutcomeDefinition): MappingResult<HttpStatusCode> = mappings.map(definition.code)

    fun map(outcome: Outcome): MappingResult<HttpStatusCode> = mappings.map(outcome.code)

    fun contains(code: OutcomeCode): Boolean = mappings.contains(code)

    fun withMapping(definition: OutcomeDefinition, status: HttpStatusCode): HttpOutcomeMapper =
        HttpOutcomeMapper(mappings.add(definition.code, status))

    fun withOverride(definition: OutcomeDefinition, status: HttpStatusCode): HttpOutcomeMapper =
        HttpOutcomeMapper(mappings.override(definition.code, status))

    fun mappings(): Map<OutcomeCode, HttpStatusCode> = mappings.entries()

    companion object {
        @JvmStatic
        fun empty(): HttpOutcomeMapper = HttpOutcomeMapper(InternalMappings.empty())

        @JvmStatic
        fun standard(): HttpOutcomeMapper =
            HttpOutcomeMapper(
                InternalMappings.from(
                    linkedMapOf(
                        StandardOutcomes.OK.code to HttpStatusCode.OK,
                        StandardOutcomes.CREATED.code to HttpStatusCode.CREATED,
                        StandardOutcomes.ACCEPTED.code to HttpStatusCode.ACCEPTED,
                        StandardOutcomes.NO_CONTENT.code to HttpStatusCode.NO_CONTENT,
                        StandardOutcomes.INVALID_ARGUMENT.code to HttpStatusCode.BAD_REQUEST,
                        StandardOutcomes.UNAUTHENTICATED.code to HttpStatusCode.UNAUTHORIZED,
                        StandardOutcomes.PERMISSION_DENIED.code to HttpStatusCode.FORBIDDEN,
                        StandardOutcomes.NOT_FOUND.code to HttpStatusCode.NOT_FOUND,
                        StandardOutcomes.ALREADY_EXISTS.code to HttpStatusCode.CONFLICT,
                        StandardOutcomes.OUT_OF_RANGE.code to HttpStatusCode.BAD_REQUEST,
                        StandardOutcomes.PAYLOAD_TOO_LARGE.code to HttpStatusCode.PAYLOAD_TOO_LARGE,
                        StandardOutcomes.RATE_LIMITED.code to HttpStatusCode.TOO_MANY_REQUESTS,
                        StandardOutcomes.UNIMPLEMENTED.code to HttpStatusCode.NOT_IMPLEMENTED,
                        StandardOutcomes.UNAVAILABLE.code to HttpStatusCode.SERVICE_UNAVAILABLE,
                        StandardOutcomes.INTERNAL.code to HttpStatusCode.INTERNAL_SERVER_ERROR,
                        StandardOutcomes.DATA_LOSS.code to HttpStatusCode.INTERNAL_SERVER_ERROR,
                    ),
                ),
            )
    }
}
